bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/processing/JsonTraceWriter.kt

```diff
@@ -0,0 +1,268 @@
+package com.bugsnag.android.performance.internal.processing
+
+import android.util.JsonWriter
+import com.bugsnag.android.performance.Logger
+import com.bugsnag.android.performance.internal.Attributes
+import com.bugsnag.android.performance.internal.SpanImpl
+import java.io.Closeable
+import java.io.Writer
+import kotlin.math.min
+
+internal class JsonTraceWriter(
+    private val json: JsonWriter,
+    private val attributeLimits: AttributeLimits? = null,
+) : Closeable by json {
+    /**
+     * We can only encode a single Span at a time - so we keep track of it here.
+     */
+    internal var currentSpan: SpanImpl? = null
+
+    constructor(out: Writer, attributeLimits: AttributeLimits? = null) :
+            this(JsonWriter(out), attributeLimits)
+
+    inline fun writeSpan(
+        spanImpl: SpanImpl,
+        scopedWrite: JsonTraceWriter.() -> Unit,
+    ) {
+        currentSpan = spanImpl
+        obj {
+            scopedWrite()
+        }
+        currentSpan = null
+    }
+
+    fun beginObject() = json.beginObject()
+
+    fun endObject() = json.endObject()
+
+    fun beginArray() = json.beginArray()
+
+    fun endArray() = json.endArray()
+
+    fun name(name: String): JsonTraceWriter {
+        json.name(name)
+        return this
+    }
+
+    fun value(value: String): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun value(value: Int): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun value(value: Long): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun value(value: Boolean): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun value(value: Double): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun value(value: Number): JsonTraceWriter {
+        json.value(value)
+        return this
+    }
+
+    fun nullValue(): JsonTraceWriter {
+        json.nullValue()
+        return this
+    }
+
+    inline fun obj(builder: JsonTraceWriter.() -> Unit): JsonTraceWriter {
+        beginObject()
+        builder()
+        endObject()
+
+        return this
+    }
+
+    inline fun array(builder: JsonTraceWriter.() -> Unit): JsonTraceWriter {
+        beginArray()
+        builder()
+        endArray()
+
+        return this
+    }
+
+    fun value(attributes: Attributes): JsonTraceWriter {
+        return array {
+            attributes.entries.forEach { (key, value) ->
+                if (isValidAttributeKey(key)) {
+                    obj {
+                        name("key").value(key)
+
+                        name("value")
+                        writeAttributeValue(key, value)
+                    }
+                } else {
+                    Logger.w("Attribute name '$key' is too long (>128)")
```

Comment:
We've defined some messages to use for these now - although it's not mandatory

Reply:
Thanks. I've updated the log messages accordingly